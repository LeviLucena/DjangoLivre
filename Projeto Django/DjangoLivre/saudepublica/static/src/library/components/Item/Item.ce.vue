<template>
  <div class="item-wrapper">
    <a
      v-if="href"
      v-tooltip="{ text: tooltipText, place: tooltipPlace, type: tooltipType, timer: tooltipTimer }"
      v-bind="$attrs"
      :href="href"
      :title="title"
      class="br-item"
      :class="{
        active: active,
        disabled: disabled,
        selected: selected,
      }"
      :data-toggle="hover && !disabled ? 'selection' : null"
    >
      <div class="content"><slot></slot></div>
    </a>

    <div
      v-else
      v-tooltip="{ text: tooltipText, place: tooltipPlace, type: tooltipType, timer: tooltipTimer }"
      v-bind="$attrs"
      class="br-item align-items-center"
      :class="{
        active: active,
        disabled: disabled,
        selected: selected,
        open: isOpen,
        'py-3': parentDensity === 'medium',
        'py-4': parentDensity === 'small',
      }"
      :role="dataToggleParentBrList ? 'listitem' : null"
      :data-toggle="hover && !disabled ? 'selection' : null"
    >
      <div v-if="dataToggleParentBrList" class="content">
        <div class="flex-fill">
          {{ title }}
          <slot name="content-header"></slot>
        </div>
        <BrButton
          circle
          type="tertiary"
          density="small"
          :icon="isOpen ? 'chevron-up' : 'chevron-down'"
          class="ml-1"
          :data-toggle="dataToggleParentBrList ? 'collapse' : hover ? 'selection' : null"
          @click="dataToggleParentBrList ? toggleOpen() : void 0"
        ></BrButton>
      </div>
      <span v-if="!dataToggleParentBrList">
        <slot></slot>
      </span>
    </div>
    <span v-show="dataToggleParentBrList && isOpen">
      <slot></slot>
    </span>
  </div>
</template>

<script>
import TooltipMixin from "../../mixins/TooltipMixin"
import BrButton from "../Button/Button.ce.vue"
import IconBase from "../IconBase/IconBase.ce.vue"

const Item = {
  name: "br-item",
  components: {
    BrButton,
    IconBase,
  },
  mixins: [TooltipMixin],
  inject: {
    parentList: {
      default: null,
    },
    parentDensity: {
      default: "small",
    },
  },
  provide() {
    return {
      parentList: this,
      parentDensity: this.parentDensity,
    }
  },
  props: {
    /**
     * **[OPCIONAL]** Coloca o item como ênfase, destacando-o dos demais.
     * */
    active: {
      type: Boolean,
      default: false,
    },
    /**
     * **[OPCIONAL]** Efeito Hover, destaca o item ao passar o mouse por cima do mesmo.
     */
    hover: {
      type: Boolean,
      default: false,
    },
    /**
     * **[OPCIONAL]** Desabilita o item, mudando o cursor do mouse para símbolo de desabilitado e oculta o item para clique.
     */
    disabled: {
      type: Boolean,
      default: false,
    },
    /**
     * **[OPCIONAL]** Destaca o item como selecionado. Possui um watcher que dispara um evento quando o seu valor é modificado.
     */
    selected: {
      type: Boolean,
      default: false,
    },
    /**
     * **[OPCIONAL]** Indica se o item está aberto/fechado quando o br-list pai é data-toggle. Possui um watcher que dispara um evento quando o seu valor é modificado.
     * */
    open: {
      type: Boolean,
      default: false,
    },
    /**
     * **[OPCIONAL]** Título visível para o item quando o br-list pai é data-toggle.
     * */
    title: {
      type: String,
      default: null,
    },
    /**
     * **[OPCIONAL]** URL ou PATH pelo qual o usuario sera encaminhado ao clicar no item. Adicionar um link torna o item do tipo link.
     * */
    href: {
      type: String,
      default: null,
    },
  },
  emits: [
    /**
     * **[OPCIONAL]** Evento disparado quando a prop *selected* é modificada.
     *
     * Referência sobre o consumo de eventos:
     * [Eventos Emitidos pelos Componentes](/?path=/story/eventos-emitidos-pelos-componentes--page)
     */
    "toggle-selected",
    /**
     * **[OPCIONAL]** Evento disparado quando a prop *open* é modificada.
     *
     * Referência sobre o consumo de eventos:
     * [Eventos Emitidos pelos Componentes](/?path=/story/eventos-emitidos-pelos-componentes--page)
     */
    "toggle-open",
  ],
  data() {
    return {
      isOpen: this.open,
    }
  },
  computed: {
    dataToggleParentBrList: function () {
      return this.parentList && this.parentList.isDataToggle()
    },

    dataUniqueParentBrList: function () {
      return this.parentList && this.parentList.isDataToggle() && this.parentList.isDataUnique()
    },
  },
  watch: {
    /**

     * Dispara o evento toggle-selected quando a prop selected é modificada.
     */
    selected: function (newValue) {
      this.$emit("toggle-selected", newValue)
    },
    /**
     * Dispara o evento toggle-open quando a prop open é modificada.
     */
    open: function (newValue) {
      this.$emit("toggle-open", newValue)
    },
  },
  mounted() {
    this.$nextTick(() => {
      if (this.dataToggleParentBrList) {
        this.parentList.addItem(this)
        this.isOpen = this.open
      } else {
        this.isOpen = true
      }
    })
  },
  methods: {
    toggleOpen() {
      if (!this.isOpen && this.dataUniqueParentBrList) {
        this.parentList.closeAllItems()
      }
      this.isOpen = !this.isOpen
      this.$emit("toggle-open", this.isOpen)
    },
  },
}

export default Item
</script>

<style lang="scss">
@import "~@govbr-ds/core/src/partial/scss/_base";
@import "~@govbr-ds/core/src/partial/scss/_utilities";
@import "~@govbr-ds/core/src/components/item/_item";
@import "~@govbr-ds/core/src/components/item/_mixins";
@import "~@govbr-ds/core/src/components/tag/_tag";
@import "~@govbr-ds/core/src/components/button/_button";
@import url("https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.min.css");

// Corrige a quebra de alinhamento no item quando o mesmo faz parte de uma lista horizontal, devido ao shadow-root.
:host {
  @include item-align-horizontal-list;
}

/* .br-item[data-toggle] {
  cursor: default;
} */
</style>
