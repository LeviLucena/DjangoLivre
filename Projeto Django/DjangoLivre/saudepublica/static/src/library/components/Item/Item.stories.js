import { action } from '@storybook/addon-actions'
import { customEventStorybookArgType, kebabiseArgs } from '../../util/Utils.js'
import BrItem from './Item.ce.vue'

import BrList from '../List/List.ce.vue'

const defaultSimpleText = `Um texto simples`

const defaultImageAndText = `
  <div class="row align-items-center">
    <div class="col-auto"><img class="rounded" src="https://picsum.photos/40/40" alt="imagem de exemplo 1" /></div>
    <div class="col"><span>Item 1 - uma imagem e um texto (sem elemento de formulário)</span></div>
  </div>
`
const checkboxItem = `
<br-checkbox id="check-01" label="Item 1"
  onchange="(function(e) {
    const checkboxItem = document.getElementById('check-01').addEventListener('onChange', (e) => {
      const clickedCheckbox = e.detail[0].checked
      const item = document.getElementById('item-01')
      clickedCheckbox ? item.setAttribute('selected', true) : item.removeAttribute('selected')
    })
  })()">
</br-checkbox>`

const itemClick = `
  (function(e) {
    const item = document.getElementById('item-01')
    item.toggleAttribute('selected')
    item.addEventListener('toggle-selected', (e) => {
      const itemSelected = e.detail[0]
      document.getElementById('check-01').shadowRoot.querySelector('input').checked = itemSelected
    })
  })()
`

export default {
  title: 'Componentes/Item',
  component: BrItem,
  BrList,
  argTypes: {
    active: {
      control: 'boolean',
      defaultValue: false,
    },
    hover: {
      control: 'boolean',
      defaultValue: false,
    },
    disabled: {
      control: 'boolean',
      defaultValue: false,
    },
    selected: {
      control: 'boolean',
      defaultValue: false,
    },
    'toggle-selected': {
      ...customEventStorybookArgType,
    },
    'toggle-open': {
      ...customEventStorybookArgType,
    },
    default: {
      control: 'text',
      defaultValue: 'Texto do item',
      description: '**[OBRIGATÓRIO]** Conteúdo do item, que deve ser passado por slot.',
      table: {
        type: {
          summary: 'string | Element',
        },
        defaultValue: {
          summary: '',
        },
      },
    },
    open: {
      defaultValue: false,
    },
    tooltipText: {
      control: 'text',
    },
    tooltipPlace: {
      options: ['top', 'right', 'bottom', 'left'],
      control: 'inline-radio',
      if: { arg: 'tooltipText', truthy: true },
    },
    tooltipType: {
      options: ['info', 'success', 'warning', 'error'],
      control: 'inline-radio',
      if: { arg: 'tooltipText', truthy: true },
    },
    tooltipTimer: {
      control: 'number',
      if: { arg: 'tooltipText', truthy: true },
    },
  },
}

const TemplateDefault = (args) => ({
  props: Object.keys(args),
  setup() {
    return { args: kebabiseArgs(args) }
  },
  template: `<br-item v-bind="args">${args.default}</br-item>`,
})

export const TextoSimples = TemplateDefault.bind({})
TextoSimples.args = {
  default: defaultSimpleText + '.',
  hover: false,
}

export const ImagemETexto = TemplateDefault.bind({})
ImagemETexto.storyName = 'Imagem e Texto'
ImagemETexto.args = {
  default: defaultImageAndText,
  hover: false,
}

export const Desabilitado = TemplateDefault.bind({})
Desabilitado.args = {
  default: defaultSimpleText + ' desabilitado.',
  disabled: true,
}

const TemplateWithCheckbox = (args) => ({
  props: Object.keys(args),
  setup() {
    return { args: kebabiseArgs(args) }
  },
  template: `
 <br-item v-bind="args"
    id="item-01"
    onclick="${args.onclick}"
    @toggle-selected="handleSelected"
    >${args.default}</br-item>`,
  methods: { handleSelected: action('toggle-selected') },
})

export const ItemDeSelecao = TemplateWithCheckbox.bind({})
ItemDeSelecao.storyName = 'Item de Seleção'
ItemDeSelecao.parameters = { controls: { exclude: ['default', 'onclick', 'selected', 'disabled'] } }
ItemDeSelecao.args = {
  default: checkboxItem,
  hover: true,
  onclick: itemClick,
}

const TemplateComList = (args) => ({
  props: Object.keys(args),
  setup() {
    return { args: kebabiseArgs(args) }
  },
  template: `<br-list title="Exemplo de item com title e open" data-toggle>
    <br-item v-bind="args" @toggle-open="handleOpen">
      <br-list>
        <br-item>Subitem 1.1</br-item>
        <br-item>Subitem 1.2</br-item>
        <br-item>Subitem 1.3</br-item>
      </br-list>
    </br-item>
  </br-list>`,
  methods: { handleOpen: action('toggle-open') },
})

export const TitleEOpen = TemplateComList.bind({})
TitleEOpen.storyName = 'Item com List: "title" e "open"'
TitleEOpen.args = {
  title: 'Título do item',
  open: true,
}

export const ComTooltip = TemplateDefault.bind({})
ComTooltip.args = {
  default: defaultSimpleText + '.',
  hover: false,
  tooltipText: 'Lorem ipsum dolor sit amet.',
}
