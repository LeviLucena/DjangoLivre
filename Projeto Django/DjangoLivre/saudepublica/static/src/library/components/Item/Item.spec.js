import { shallowMount } from '@vue/test-utils'
import BrDivider from '../Divider/Divider.ce.vue'
import BrList from '../List/List.ce.vue'
import Item from './Item.ce.vue'

describe('br-item', () => {
  test('Testa se a prop active está renderizando.', () => {
    const wrapper = shallowMount(Item, {
      propsData: { active: true },
    })
    expect(wrapper.find('.active').exists()).toBe(true)
  })

  test('Testa se a prop disabled está renderizando', () => {
    const wrapper = shallowMount(Item, {
      propsData: { disabled: true },
    })
    expect(wrapper.find('.disabled').exists()).toBe(true)
  })

  test('Testa se a prop selected está renderizando.', () => {
    const wrapper = shallowMount(Item, {
      propsData: { selected: true },
    })
    expect(wrapper.find('.selected').exists()).toBe(true)
  })

  test('Testa se o evento toggle-selected foi disparado', async () => {
    const wrapper = shallowMount(Item, {
      propsData: { selected: false },
    })
    wrapper.setProps({ selected: true })
    wrapper.vm.$emit('toggle-selected')
    await wrapper.vm.$nextTick()
    expect(wrapper.emitted('toggle-selected')).toBeTruthy()
  })

  test('Testa se o evento toggle-open foi disparado', async () => {
    const wrapper = shallowMount(Item, {
      propsData: { open: false },
    })
    wrapper.setProps({ open: true })
    wrapper.vm.$emit('toggle-open')
    await wrapper.vm.$nextTick()
    expect(wrapper.emitted('toggle-open')).toBeTruthy()
  })

  test('has the expected html structure', () => {
    const wrapper = shallowMount(Item)
    expect(wrapper.element).toMatchSnapshot()
  })

  const listItens = `
    <br-item title="Item 1">
      <br-list>
        <br-item>Subitem 1.1</br-item>
        <br-item>Subitem 1.2</br-item>
        <br-item>Subitem 1.3</br-item>
      </br-list>
    </br-item>

    <br-divider></br-divider>
    <br-item title="Item 2">
      <br-list>
        <br-item>Detalhes do item 2</br-item>
      </br-list>
    </br-item>
  `

  test('insert item into list', () => {
    const wrapper = shallowMount(BrList, {
      propsData: {
        dataToggle: true,
        dataUnique: true,
      },
      slots: {
        default: listItens,
      },
      global: {
        stubs: {
          'br-item': Item,
          'br-list': BrList,
          'br-divider': BrDivider,
        },
      },
    })
    expect(wrapper.findComponent(Item))
    expect(wrapper.element).toMatchSnapshot()
  })

  test('Testa se o valor de dataToggleParentBrList para List com propriedade data-toggle igual a true', () => {
    const wrapper = shallowMount(BrList, {
      propsData: {
        dataToggle: true,
      },
      slots: {
        default: listItens,
      },
      global: {
        stubs: {
          'br-item': Item,
          'br-list': BrList,
          'br-divider': BrDivider,
        },
      },
    })
    const item = wrapper.findComponent(Item)
    expect(item.vm.dataToggleParentBrList).toBeTruthy()
  })

  test('Testa se o valor de dataUniqueParentBrList para List com propriedade data-unique igual a true', () => {
    const wrapper = shallowMount(BrList, {
      propsData: {
        dataToggle: true,
        dataUnique: true,
      },
      slots: {
        default: listItens,
      },
      global: {
        stubs: {
          'br-item': Item,
          'br-list': BrList,
          'br-divider': BrDivider,
        },
      },
    })
    const list = wrapper.findComponent(Item)
    expect(list.vm.dataUniqueParentBrList).toBeTruthy()
  })

  test('Testa se o método dataToggle foi chamado', () => {
    const wrapper = shallowMount(Item, {
      propsData: { open: false },
    })
    const item = wrapper.find('.br-item')
    item.trigger('click')
    wrapper.vm.toggleOpen()
  })
})
