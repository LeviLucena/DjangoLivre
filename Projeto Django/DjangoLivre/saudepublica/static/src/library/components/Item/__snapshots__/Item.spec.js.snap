// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`br-item has the expected html structure 1`] = `
<div
  class="item-wrapper"
>
  <div
    class="br-item align-items-center py-4"
  >
    <!--v-if-->
    <span>
      
      
    </span>
  </div>
  <span
    style="display: none;"
  >
    
    
  </span>
</div>
`;

exports[`br-item insert item into list 1`] = `
<div
  class="br-list"
  data-toggle="true"
  data-unique="true"
  density="large"
  role="list"
>
  <!--v-if-->
  <!--v-if-->
  
  <div
    class="item-wrapper"
  >
    <div
      class="br-item align-items-center"
      role="listitem"
    >
      <div
        class="content"
      >
        <div
          class="flex-fill"
        >
          Item 1 
          
          
        </div>
        <br-button-stub
          active="false"
          block="false"
          circle="true"
          class="ml-1"
          data-toggle="collapse"
          density="small"
          disabled="false"
          icon="chevron-down"
          inverted="false"
          loading="false"
          type="tertiary"
        />
      </div>
      <!--v-if-->
    </div>
    <span
      style="display: none;"
    >
      
      <div
        class="br-list"
        data-toggle="false"
        data-unique="false"
        density="large"
        role="list"
      >
        <!--v-if-->
        <!--v-if-->
        
        <div
          class="item-wrapper"
        >
          <div
            class="br-item align-items-center"
          >
            <!--v-if-->
            <span>
              
              Subitem 1.1
              
            </span>
          </div>
          <span
            style="display: none;"
          >
            
            Subitem 1.1
            
          </span>
        </div>
        <div
          class="item-wrapper"
        >
          <div
            class="br-item align-items-center"
          >
            <!--v-if-->
            <span>
              
              Subitem 1.2
              
            </span>
          </div>
          <span
            style="display: none;"
          >
            
            Subitem 1.2
            
          </span>
        </div>
        <div
          class="item-wrapper"
        >
          <div
            class="br-item align-items-center"
          >
            <!--v-if-->
            <span>
              
              Subitem 1.3
              
            </span>
          </div>
          <span
            style="display: none;"
          >
            
            Subitem 1.3
            
          </span>
        </div>
        
      </div>
      
    </span>
  </div>
  <span
    class="br-divider"
  />
  <div
    class="item-wrapper"
  >
    <div
      class="br-item align-items-center"
      role="listitem"
    >
      <div
        class="content"
      >
        <div
          class="flex-fill"
        >
          Item 2 
          
          
        </div>
        <br-button-stub
          active="false"
          block="false"
          circle="true"
          class="ml-1"
          data-toggle="collapse"
          density="small"
          disabled="false"
          icon="chevron-down"
          inverted="false"
          loading="false"
          type="tertiary"
        />
      </div>
      <!--v-if-->
    </div>
    <span
      style="display: none;"
    >
      
      <div
        class="br-list"
        data-toggle="false"
        data-unique="false"
        density="large"
        role="list"
      >
        <!--v-if-->
        <!--v-if-->
        
        <div
          class="item-wrapper"
        >
          <div
            class="br-item align-items-center"
          >
            <!--v-if-->
            <span>
              
              Detalhes do item 2
              
            </span>
          </div>
          <span
            style="display: none;"
          >
            
            Detalhes do item 2
            
          </span>
        </div>
        
      </div>
      
    </span>
  </div>
  
</div>
`;
